port/unix/omrsignal.c

```diff
@@ -1450,30 +1509,33 @@ registerMasterHandlers(OMRPortLibrary *portLibrary, uint32_t flags, uint32_t all
 		return OMRPORT_SIG_ERROR;
 	}
 
-	if (0 != flagsSignalsOnly) {
-		/* registering some handlers */
-		uint32_t portSignalType = 0;
+	/* Check if any signal bits are set in flagsSignalsOnly by masking the control bits. */
+	if (0 != (flagsSignalsOnly & ~OMRPORT_SIG_FLAG_CONTROL_BITS_MASK)) {
+		/* Register handlers. */
+		uint32_t portSignalFlag = 0;
 
 		/* OMRPORT_SIG_SMALLEST_SIGNAL_FLAG represents the smallest signal
-		 * flag. portSignalType is initialized to the smallest signal flag
+		 * flag. portSignalFlag is initialized to the smallest signal flag
 		 * in order to avoid non-signal flags. Any non-signal flags greater
 		 * than the smallest signal flag are ignored via a bitwise-and
 		 * operation with allowedSubsetOfFlags. allowedSubsetOfFlags either
 		 * represents all synchronous signal flags (OMRPORT_SIG_FLAG_SIGALLSYNC)
 		 * or all asynchronous signal flags (OMRPORT_SIG_FLAG_SIGALLASYNC).
 		 */
-		for (portSignalType = OMRPORT_SIG_SMALLEST_SIGNAL_FLAG; ((portSignalType < allowedSubsetOfFlags) && (portSignalType != 0)); portSignalType = portSignalType << 1) {
-			/* iterate through all the  signals and register the master handler for those that don't have one yet */
-
-			if (OMR_ARE_ALL_BITS_SET(flagsSignalsOnly, portSignalType)) {
-				if (OMR_ARE_NO_BITS_SET(signalsWithMasterHandlers, portSignalType)) {
-					/* Register a master handler for this (portSignalType's) signal. */
-					if (0 != registerSignalHandlerWithOS(portLibrary, portSignalType, handler, oldOSHandler)) {
+		for (portSignalFlag = OMRPORT_SIG_SMALLEST_SIGNAL_FLAG; ((portSignalFlag < allowedSubsetOfFlags) && (portSignalFlag != 0)); portSignalFlag = portSignalFlag << 1) {
```

Comment:
This loop will have trouble if the most significant bit is used for some supported signal.
It should be replaced with this structure:
```
uint32_t signalFlags = flagsSignalsOnly & ~OMRPORT_SIG_FLAG_CONTROL_BITS_MASK;

while (0 != signalFlags) {	
	uint32_t smallestSignal = signalFlags & -signalFlags;
	/* process signal corresponding to smallestSignal */
	signalFlags ^= smallestSignal;
}
```
An added bonus is that it only iterates over the signals actually specified rather than all possible signals.

Reply:
amended.